### PR TITLE
Implement create_operator_func API.

### PR DESCRIPTION
**Pull request by lyupan**
pgtle_admin can run `create_operator_func` to create operator funcs for base types created in pg_tle (by `create_base_type`).

User-defined operator function must accept one or two arguments of BYTEA, we make an C-version operator function that accepts base type as the argument; inside the C-version operator function, we find the user-defined operator function and execute it.

Notes:
1. `create_operator_func` is not strictly necessary to define an operator function on the base type; however, it will be required for languages such as plrust, because the custom base type is not available in plrust;
2. `CREATE OPERATOR CLASS` requires superuser privileges in pg_tle, but should be already supported in AWS RDS.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/datatype.c [hunk 1]
```diff
@@ -704,3 +707,271 @@ get_qualified_funcname(Oid funcid)
...
+		ReleaseSysCache(tuple);
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
+				 errmsg("type operator functions must exist in the same namespace as the type")));
+	}
+
+	argTypes = (Oid *) palloc(proc->pronargs * sizeof(Oid));
+	argTypes[0] = typeOid;
+	if (proc->pronargs == 2)
+		argTypes[1] = typeOid;
```
Comment (jkatz) [inline src/datatype.c:841]:
```suggestion
	if (proc->pronamespace != expectedNamespace)
	{
		ReleaseSysCache(tuple);
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
				 errmsg("type operator functions must exist in the same namespace as the type")));
	}

	if (proc->pronargs > 2)
	{
		ReleaseSysCache(tuple);
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
				 errmsg("type opeartor function must accept one or two arguments bytea")));
	}

	argTypes = (Oid *) palloc(proc->pronargs * sizeof(Oid));

	/* FIXME: declare int i at the top of the function */
	for (i = 0; i < proc->pronargs; i++)
	{
		if (proc->proargtypes.values[i] != BYTEAOID)
		{
			ReleaseSysCache(tuple);
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
					errmsg("type operator function must accept arguments of bytea")));	
		}

		argTypes[i] = typeOid;
	}
```

File: src/datatype.c [hunk 2]
```diff
@@ -704,3 +707,271 @@ get_qualified_funcname(Oid funcid)
...
+
+	funcNameList = list_make2(makeString(get_namespace_name(expectedNamespace)),
+							  makeString(NameStr(proc->proname)));
+	if (OidIsValid(LookupFuncName(funcNameList, proc->pronargs, argTypes, true)))
+		ereport(ERROR,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("function \"%s\" already exists", NameListToString(funcNameList))));
```
Comment (jkatz) [inline src/datatype.c:848]: This doesn't release the tuple.

Also -- are we able to release the tuple after we copy the data into the structure? I believe we'd have to do a full copy of the tuples.

**Review by jkatz — COMMENTED** [review]
Just a few comments to get started with. Need to dive deeper into this.

File: src/datatype.c (hunk 1 above)
Comment (lyupan) [reply src/datatype.c]: Updated. Thanks for the suggestion!

File: src/datatype.c (hunk 2 above)
Comment (lyupan) [reply src/datatype.c]: > This doesn't release the tuple

Good catch! Fixed in a new commit.

> Also -- are we able to release the tuple after we copy the data into the structure? I believe we'd have to do a full copy of the tuples.

Do you refer to this code `makeString(NameStr(proc->proname))`? That's probably an issue, I've added a `pstrdup` to do a string copy.

File: src/datatype.c [hunk 3]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
 
 	proc = (Form_pg_proc) GETSTRUCT(tuple);
 	inputNames = list_make2(makeString(get_namespace_name(proc->pronamespace)),
-							makeString(NameStr(proc->proname)));
+							makeString(pstrdup(NameStr(proc->proname))));
```
Comment (jkatz) [inline src/datatype.c:706]: This is pretty hard to read. Likewise, if we need to make a copy of `proc->proname`, we'd also have to do that for `proc->pronamespace`  (unless `get_namespace_name` is doing that).

File: src/datatype.c [hunk 4]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+
+	/*
+	 * check_user_operator_func already ensures the number of func args is 1
+	 * or 2.
+	 */
+	nargs = get_func_nargs(funcOid);
+	argTypes = (Oid *) palloc(nargs * sizeof(Oid));
+	argTypes[0] = typeOid;
+	if (nargs == 2)
+		argTypes[1] = typeOid;
```
Comment (jkatz) [inline src/datatype.c:780]:
```suggestion
	for (i = 0; i < nargs; i++)
		argTypes[i] = typeOid;	
```

File: src/datatype.c [hunk 5]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+
+	tleadminoid = get_role_oid(PG_TLE_ADMIN, false);
+	typeOwner = typeForm->typowner;
+	inputOid = typeForm->typinput;
+	outputOid = typeForm->typoutput;
+	ReleaseSysCache(tuple);
+
+	CHECK_CAN_SET_ROLE(typeOwner, tleadminoid);
+
+	if (!is_pgtle_io_func(inputOid, true) || !is_pgtle_io_func(outputOid, false))
```
Comment (jkatz) [inline src/datatype.c:899]:
```suggestion
	if (!(is_pgtle_io_func(inputOid, true) && is_pgtle_io_func(outputOid, false)))
```

File: src/datatype.c [hunk 6]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+		return false;
+	}
+
+	prosrcattr = SysCacheGetAttr(PROCOID, tuple,
+								 Anum_pg_proc_prosrc, &isnull);
+	Assert(!isnull);
+
+	prosrcstring = TextDatumGetCString(prosrcattr);
+	ReleaseSysCache(tuple);
+	return strcmp(prosrcstring, typeInput ? TLE_BASE_TYPE_IN : TLE_BASE_TYPE_OUT) == 0;
```
Comment (jkatz) [inline src/datatype.c:942]: We've been using `strncmp`

File: src/datatype.c [hunk 7]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+	if (proc->pronamespace != expectedNamespace)
+	{
+		ReleaseSysCache(tuple);
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
+				 errmsg("type operator functions must exist in the same namespace as the type")));
+	}
+
+	nargs = proc->pronargs;
+	if (nargs != 1 && nargs != 2)
```
Comment (jkatz) [inline src/datatype.c:832]:
```suggestion
	if (nargs < 1 || nargs > 2)
```

File: src/datatype.c [hunk 8]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+{
+	Datum		result;
+	Oid			userFunc;
+	List	   *procname;
+	Oid		   *argtypes = NULL;
+	int			nargs = 0;
+
+	procname = get_qualified_funcname(fcinfo->flinfo->fn_oid);
+	get_func_signature(fcinfo->flinfo->fn_oid, &argtypes, &nargs);
+	if (nargs != 1 && nargs != 2)
```
Comment (jkatz) [inline src/datatype.c:963]:
```suggestion
	if (nargs < 1 || nargs > 2)
```

File: src/datatype.c [hunk 9]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
+				 errmsg("operator function %s must accept one or two arguments",
+						func_signature_string(procname, nargs, NIL, argtypes))));
+
+	argtypes[0] = BYTEAOID;
+	if (nargs == 2)
+		argtypes[1] = BYTEAOID;
```
Comment (jkatz) [inline src/datatype.c:971]:
```suggestion
	for (i = 0; i < 2; i++)
		argtypes[i] = BYTEAOID;
```

File: src/datatype.c [hunk 10]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				 errmsg("function %s does not exist",
+						func_signature_string(procname, nargs, NIL, argtypes))));
+
+	if (nargs == 1)
+		result = OidFunctionCall1Coll(userFunc, InvalidOid, PG_GETARG_DATUM(0));
+	else
+		result = OidFunctionCall2Coll(userFunc, InvalidOid, PG_GETARG_DATUM(0), PG_GETARG_DATUM(1));
+
+	return result;
```
Comment (jkatz) [inline src/datatype.c:985]:
```suggestion
	if (nargs == 1)
		return OidFunctionCall1Coll(userFunc, InvalidOid, PG_GETARG_DATUM(0));

	return  OidFunctionCall2Coll(userFunc, InvalidOid, PG_GETARG_DATUM(0), PG_GETARG_DATUM(1));
```

**Review by jkatz — COMMENTED** [review]
I think this is just about ready. Most of my comments were minor.

File: src/datatype.c [hunk 11]
```diff
@@ -700,7 +703,284 @@ get_qualified_funcname(Oid funcid)
...
+	Oid		   *argTypes;
+	AclResult	aclresult;
+	char	   *namespaceName;
+
+	/*
+	 * Even though the SQL function is locked down so only a member of
+	 * pgtle_admin can run this function, let's check and make sure there is
+	 * not a way to bypass that
+	 */
+	check_is_pgtle_admin();
```
Comment (formalLogicGirl) [inline src/datatype.c:737]: Can we add some regression tests for the various checks we exercise in this function? Such as
* a non pgtle_admin role attempts to call the function -- fails
* role without CREATE permissions in namespace attempts to call the function -- fails
* role is not owner of base type -- fails
* role is not owner of operator func -- fails

File: src/datatype.c (hunk 4 above)
Comment (lyupan) [reply src/datatype.c]: Done. Thanks for the suggestion

File: src/datatype.c (hunk 5 above)
Comment (lyupan) [reply src/datatype.c]: Done

File: src/datatype.c (hunk 6 above)
Comment (lyupan) [reply src/datatype.c]: Done

File: src/datatype.c (hunk 7 above)
Comment (lyupan) [reply src/datatype.c]: Done

File: src/datatype.c (hunk 8 above)
Comment (lyupan) [reply src/datatype.c]: Done

File: src/datatype.c (hunk 9 above)
Comment (lyupan) [reply src/datatype.c]: Done

File: src/datatype.c (hunk 10 above)
Comment (lyupan) [reply src/datatype.c]: Done

File: src/datatype.c (hunk 11 above)
Comment (lyupan) [reply src/datatype.c:742]: Done. Added a few more regression tests cases.

File: src/datatype.c [hunk 12]
```diff
@@ -693,14 +696,294 @@ get_qualified_funcname(Oid funcid)
...
+	char	   *expectedProsrc;
+
+	tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(funcid));
+	if (!HeapTupleIsValid(tuple))
+		elog(ERROR, "cache lookup failed for function %u", funcid);
+
+	proc = (Form_pg_proc) GETSTRUCT(tuple);
+	if (proc->prolang != ClanguageId)
+	{
+		ReleaseSysCache(tuple);
```
Comment (formalLogicGirl) [inline src/datatype.c:937]: Similar comment as previous one, about copying the data out, and once only ReleaseSysCache(tuple)

Comment (lyupan) [reply src/datatype.c:938]: `SysCacheGetAttr` is somewhat more costly than a simply copy, and there are only two `ReleaseSysCache`, I guess we can keep it as it is.

**Review by formalLogicGirl — APPROVED** [review]
Looks good to me!

**Pull request merged**